### PR TITLE
414 positioning solver

### PR DESCRIPTION
**Pull request by chris-jenkins-isi**
In progress:
- Add orientation to BoundingBox 
- Test collision detection in non-axis-aligned situations (but no edge-to-edge collisions)
- Convert to matrix representations, configure for solving via linear programming

### CODE REVIEW

File: adam/visualization/positioning_test.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    print(
+        f"test dot for corner . axis: {test_dot.shape}, type: {type(test_dot)}, value: {test_dot}"
+    )
+
+    projections = [
+        np.dot(resolved_corners[j], relative_bb_faces[i])
+        for i in range(3)
+        for j in range(8)
+    ]
+    print(f"shape of projection: {projections[0].shape}")
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning_test.py]: This shape is being reported as a vector3 (3,)

The desired operation is the dot product of a _particular_ corner (3,) and one of the projecting axes of the stationary BoundingBox (another (3,) ), which should result in a scalar.

File: adam/visualization/positioning_test.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        f"{relative_bb_min_max_projections[0]}, type: {type(relative_bb_min_max_projections[0])}"
+    )
+
+    print(f"type of projections[0][0]: {type(projections[0][0])}")
+    print(
+        f"type of relative_bb_min_max_projections[0]: {type(relative_bb_min_max_projections[0])}"
+    )
+
+    # This is very wrong because we want to disincentivize collisions
+    collide_constraint = (
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning_test.py]: What's happening here is an attempt to test a non-collision constraint between _one particular corner's projection_, and the precomputed min/max projection from the stationary Box. 

This is the disjunction issue again, in the sense that there are two ways in which an interval can not intersect with another. 

Currently the inequality is a `>=` due to the curvature of the functions on its left hand side. (The whole thing: minimum() - maximum() comes out as concave).

File: adam/visualization/positioning.py [hunk 3]
```diff
@@ -33,86 +33,156 @@ def __init__(self, main_point: ndarray, points: List[ndarray]) -> None:
...
+        )
+
+        print(f"shape of overlap distance: {overlap_distance.shape}")  # should be (3, 1)
+        print(f"overlap distance: {overlap_distance}")
+
+        # if ANY element in overlap_distance is positive, the minimum positive value
+        # is the separation distance
+
+        # otherwise the penetration distance is the maximum negative value
+        # (the smallest translation that would disentangle the two
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning.py]: Here is the tricky part -- I'm not sure how to express the `if/else` sort of criteria here.

File: adam/visualization/positioning.py [hunk 4]
```diff
@@ -44,99 +46,195 @@ def create_at_random_position(
...
-        print(f"face min-maxes: {face_min_max}")
+        Returns:
+            (3, 8) tensor -> each point projected onto each of three dimensions
 
-        # compute overlap between the parameter's min/maxes and the static ones
+        """
+        check_arg(axes.shape == (3,3))
+        corners = self.get_corners()
+        return torch.tensor([
+            corners[j].dot(axes[i]) for i in range(3) for j in range(8)
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning.py]: Turns out I was wrong about which section is tricky! I need to get rid of the _iterating over one tensor to produce another_ code in three of my functions s/t a gradient can be computed all the way back to `AxisAlignedBoundingBox.center`

File: adam/visualization/positioning.py [hunk 5]
```diff
@@ -44,99 +46,195 @@ def create_at_random_position(
...
                     torch.max(
-                        face_min_max[i][0], self.static_bb_min_max_projections[i][0]
+                        min_max_proj_0[i][0], min_max_proj_1[i][0]
                     ),
                     torch.min(
-                        face_min_max[i][1], self.static_bb_min_max_projections[i][1]
+                        min_max_proj_0[i][1], min_max_proj_1[i][1]
                     ),
                 ]
                 for i in range(3)
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning.py]: Another iterative operation creating a new tensor

File: adam/visualization/positioning.py [hunk 6]
```diff
@@ -44,99 +46,195 @@ def create_at_random_position(
...
+
+        Returns: Tensor with a scalar of the collision penalty size
+
+        """
+        # print(f"overlap penalty: min_max_overlaps shape: {min_max_overlaps.shape}")
 
         overlap_distance = torch.tensor(
-            [overlap_ranges[i][0] - overlap_ranges[i][1] for i in range(3)],
-            dtype=torch.float,
+            [min_max_overlaps[i][0] - min_max_overlaps[i][1] for i in range(3)],
```
Comment (chris-jenkins-isi) [inline adam/visualization/positioning.py]: The last such operation

File: adam/visualization/positioning.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+        for dim in range(3):
+            if 0 > overlap_distance[dim] >= overlap_distance[smallest_colliding_dim]:
+                smallest_colliding_dim = dim
+
+        return overlap_distance[smallest_colliding_dim] * -10 + 1
+
+
+class PositionSolver:
+    # constants to eventually use
+    POSITION_CONSTRAINTS = {
```
Comment (gabbard) [inline adam/visualization/positioning.py]: What do these various constraints mean?

File: adam/visualization/positioning.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        "X_RANGE": (-10.0, 10.0),
+        "Y_RANGE": (-10.0, 5.0),
+        "Z_RANGE_GROUND": (0.0, 0.0),
+        "Z_RANGE_FLY": (2.0, 5.0),
+        "Z_RANGE_OTHER": (0.0, 3.0),
+    }
+
+    CLOSE_DISTANCE = 0.25
```
Comment (gabbard) [inline adam/visualization/positioning.py]: What is the constant for?

File: adam/visualization/positioning.py [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+    def __init__(self, main_point: ndarray, points: List[ndarray]) -> None:
```
Comment (gabbard) [inline adam/visualization/positioning.py]: What do `main_point` and `points` mean?

File: adam/visualization/positioning.py [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    def forward(
+        self,
+        bounding_box_1: AxisAlignedBoundingBox,
+        bounding_box_2: AxisAlignedBoundingBox,
+    ) -> torch.Tensor:
+
+        # get face norms from one of the boxes:
+        face_norms = bounding_box_2.get_face_norms()
+
+        return AxisAlignedBoundingBox.overlap_penalty(
```
Comment (gabbard) [inline adam/visualization/positioning.py]: I would shift all the overlap logic into this class, rather than on `AxisAlignedBoundingBox`

Comment (gabbard) [reply adam/visualization/positioning.py]: Also this is complicated enough to need some explanation (which could be a reference to somewhere else which explains it in detail)

File: adam/visualization/positioning.py [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+
+        smallest_colliding_dim = 0
+        for dim in range(3):
+            if 0 > overlap_distance[dim] >= overlap_distance[smallest_colliding_dim]:
+                smallest_colliding_dim = dim
+
+        return overlap_distance[smallest_colliding_dim] * -10 + 1
+
+
+class PositionSolver:
```
Comment (gabbard) [inline adam/visualization/positioning.py]: 📖

File: adam/visualization/positioning.py [hunk 12]
```diff
@@ -0,0 +1,371 @@
+from itertools import combinations
```
Comment (gabbard) [inline adam/visualization/positioning.py]: Try to rearrange the things in this module in the order you'd like a reader to encounter them (as much as Python will allow)

File: adam/visualization/make_scenes.py [hunk 13]
```diff
@@ -22,46 +28,89 @@
...
+    return render_obj_nested(renderer, properties, obj, None)
+
+
+def render_obj_nested(
+    renderer: SituationVisualizer,
+    properties: DefaultDict[
+        ObjectPerception, List[Optional[Union[RgbColorPerception, OntologyNode]]]
+    ],
+    obj: ObjectPerception,
+    parent: NodePath,
```
Comment (gabbard) [inline adam/visualization/make_scenes.py]: Shouldn't the type annotation here be `Optional[NodePath]` if we are passing `None` in `render_obj`

File: adam/visualization/make_scenes.py [hunk 14]
```diff
@@ -22,46 +28,89 @@
...
+        viz.print_scene_graph()
         input("Press ENTER to continue")
         viz.clear_scene()
         viz.run_for_seconds(0.25)
 
 
+def render_obj(
+    renderer: SituationVisualizer,
+    properties: DefaultDict[
+        ObjectPerception, List[Optional[Union[RgbColorPerception, OntologyNode]]]
```
Comment (gabbard) [inline adam/visualization/make_scenes.py]: nit: A type alias for the union may make things clearer.

File: adam/visualization/make_scenes.py [hunk 15]
```diff
@@ -22,46 +28,89 @@
...
+    obj: ObjectPerception,
+    parent: NodePath,
+) -> NodePath:
+
+    if obj.geon is None:
+        return renderer.add_dummy_node(obj.debug_handle, parent)
+    shape = SceneCreator.cross_section_to_geo(obj.geon.cross_section)
+    # TODO***: allow for Irregular geons to be rendered
+    if shape == Shape.IRREGULAR:
+        return
```
Comment (gabbard) [inline adam/visualization/make_scenes.py]: We should at least log a warning and possibly crash.

File: adam/visualization/make_scenes.py [hunk 16]
```diff
@@ -3,11 +3,17 @@
    operating and when other code (gathering and processing scene information)
    is executing in a serial manner.
    """
-from typing import Iterable, List, Tuple, Union, DefaultDict
+from typing import Iterable, List, Tuple, Union, DefaultDict, Optional, Callable, Any
```
Comment (gabbard) [inline adam/visualization/make_scenes.py]: If this module is not yet using the positioning stuff, can we shift it to another PR that I can review later?
